caffe2/python/optimizer.py

```diff
@@ -185,6 +184,54 @@ def scale_learning_rate(self, scale):
         self.base_learning_rate *= scale
         return
 
+class MultiPrecisionSgdOptimizer(SgdOptimizer):
+    def __init__(self, base_learning_rate, momentum, policy, nesterov=1, **kwargs):
+        super(SgdOptimizer, self).__init__()
+        self.base_learning_rate = base_learning_rate
+        self.momentum = momentum
+        self.policy = policy
+        self.nesterov = nesterov
+        self.init_kwargs = kwargs
+
+    def _run(self, net, param_init_net, param_info):
+        param = param_info.blob
+        param_fp32 = param_info.blob_copy['float'] if param_info.blob_copy is not None else None
+
+        # If we have a straight fp32 parameter, run the base class
+        if param_fp32 == None:
+            return SgdOptimizer._run(self, net, param_init_net, param_info)
+
+        grad = param_info.grad
+        if self.base_learning_rate <= 0:
+            return
+
+        lr, _ = self.build_lr(
+            net, param_init_net,
+            base_learning_rate=-self.base_learning_rate,
+            learning_rate_blob=param + "_lr",
+            policy=self.policy,
+            **(self.init_kwargs)
+        )
+
+        momentum_data = param_init_net.ConstantFill(
+            param_fp32, str(param) + "_momentum", value=0.)
+        self._aux_params.local.append(momentum_data)
+
+        assert not isinstance(grad, core.GradientSlice), \
+                "Doesn't support sparse gradients"
+
+        # Copy gradient to fp32
+        grad_fp32 = net.HalfToFloat(grad, grad + "_fp32")
+
+        # update (fused) in fp32
+        net.MomentumSGDUpdate(
```

Comment:
Not suggesting to do it in this PR, but going forward it would be nice to have a way to convert an optimizer to fp16 format without replicating its  code.  What do you think?